baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/LoggingDoNotLog.java

```diff
@@ -0,0 +1,119 @@
+import com.palantir.baseline.errorprone.safety.SafetyAnalysis;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.NewClassTree;
+import com.sun.source.util.TreePath;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Ensures that data which has been marked as {@code @DoNotLog} is not passed to a logger, or an exceptoin which
+ * will almost certainly be passed to a logger.
+ */
+@AutoService(BugChecker.class)
+@BugPattern(
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = BugPattern.LinkType.CUSTOM,
+        severity = BugPattern.SeverityLevel.ERROR,
+        summary = "@DoNotLog types must not be passed to any logger directly or indirectly, for example respectively: "
+                + "log.info(doNotLog) or throw new RuntimeException(doNotLog.toString()). Exceptions are almost "
+                + "always passed to a logger in some form, and must not include data that cannot be logged.")
+public final class LoggingDoNotLog extends BugChecker
+        implements BugChecker.MethodInvocationTreeMatcher, BugChecker.NewClassTreeMatcher {
+
+    private static final Matcher<ExpressionTree> THROWABLE_CTOR = MethodMatchers.constructor()
+            .forClass(TypePredicates.allOf(
+                    TypePredicates.isDescendantOf(Throwable.class.getName()),
+                    // Avoid double-checking safe-loggable implementations which are handled by
+                    // IllegalSafeLoggingArgument
+                    TypePredicates.not(TypePredicates.isDescendantOf("com.palantir.logsafe.SafeLoggable"))));
+
+    // Note that we don't check SafeLogger here because it is handled by the IllegalSafeLoggingArgument check.
+    private static final Matcher<ExpressionTree> LOGGING_METHODS = Matchers.anyOf(
+            MethodMatchers.instanceMethod().onDescendantOf("org.slf4j.Logger"),
+            MethodMatchers.instanceMethod().onDescendantOf("org.apache.log4j.Logger"),
+            MethodMatchers.instanceMethod().onDescendantOf("org.apache.logging.log4j.Logger"),
+            MethodMatchers.instanceMethod()
+                    .onDescendantOf(System.Logger.class.getName())
+                    .named("log"),
+            MethodMatchers.instanceMethod().onDescendantOf("java.util.logging.Logger"),
+            // MDC interactions result in data in the logs
+            MethodMatchers.staticMethod().onClass("org.slf4j.MDC"),
+            MethodMatchers.staticMethod().onClass("org.apache.log4j.MDC"),
+            MethodMatchers.staticMethod().onClass("org.apache.logging.log4j.ThreadContext"));
+
+    // Result in indirect logging
+    private static final Matcher<ExpressionTree> PRECONDITIONS_METHODS = Matchers.anyOf(
+            MethodMatchers.staticMethod()
+                    .onClass("com.google.common.base.Preconditions")
+                    .namedAnyOf("checkArgument", "checkNotNull", "checkState"),
```

Comment:
Should this just be any method? E.g. there is also `checkElementIndex` and `checkPositionIndex`.

Reply:
I tried to cover the most common methods, supporting others requires a different argument index offset. Could be a future improvement to this check.